TempPath (<: AutoCloseable) and withTempFile|Dir convenience methods

mpollmeier opened the pull request:
Implements https://github.com/com-lihaoyi/os-lib/discussions/140

`TempPath` implements `AutoCloseable` so we can use automatic resource management (e.g. `Using`) to automatically remove these temporary resources once they're not needed any longer. 

We no longer rely on the JVM's `deleteOnExit` handler, therefor the temp file gets deleted even if JVM never stopps gracefully (e.g. because it's intended to run continually, or the process get's killed from the outside).

`withTempFile` and `withTempDir` are convenience methods that create a temporary file or directory, passes them to a given function and remove them immediately after the function completed, even if the function threw an exception.

```scala
----
withTempFile { file =>
  os.write(file, "some content")
}

withTempDir { file =>
  val file = dir / "somefile"
  os.write(file, "some content")
}
```

lefou reviewed (commented):
It looks almost good, but I'm a bit undecided about the fact, that we now always return `TempPath`s and still provide the `deleteOnExit` option. Something in me calls for a cleaner approach, e.g. unregister the delete-on-exit shutdown hook after we already closed the `TempFile`. There is also this other thought, that deleting the file/dir might fail (maybe because the resource is locked) and we could fall-back to the deleteOnExit strategy once we know we failed in the `close` method. But these thoughts are still a bit wishy-washy, so let me overthink it a bit longer and tell my what you think.

lefou commented:
Also, shouldn't we in `withTempFile` and `withTempDir` provide the same set of options as we do in the `apply` methods?

mpollmeier commented on os/src/TempOps.scala:
```diff
@@ -67,7 +79,38 @@ object temp {
+    * }
+    * }}} 
+    */
+  def withTempFile[A](fun: Path => A): Try[A] =
```
agreed, done

mpollmeier commented on os/src/TempOps.scala:
```diff
@@ -67,7 +79,38 @@ object temp {
+    * }
+    * }}} 
+    */
+  def withTempDir[A](fun: Path => A): Try[A] =
```
agreed, done

mpollmeier commented:
> shouldn't we in withTempFile and withTempDir provide the same set of options as we do in the apply methods?

Done as well

mpollmeier commented:
Re your other thoughts with "TempPaths and still provide the deleteOnExit option" etc. - I don't have a strong opinion either way, but would prefer to keep it simple. Automagically unregistering the deletion sounds like opening a can of worms, and we may get surprised by how many edge cases something as simple as this can give us...

mpollmeier commented:
Re supporting scala versions - does mill have standard src dirs like sbt's `src/main/scala-2.13` and `src/main/scala-3` etc? Do you have an idea on how to handle that best for this build?

lefou commented:
> Re supporting scala versions - does mill have standard src dirs like sbt's `src/main/scala-2.13` and `src/main/scala-3` etc? Do you have an idea on how to handle that best for this build?

Yeah, Mill's `CrossScalaModule` does support these source dirs by default, but in this build, the `sources` is already overridden, to support platform specific sources like `src-jvm`. 

https://github.com/com-lihaoyi/os-lib/blob/224aae739fa861b4b4302dd8a7c0204536f38e91/build.sc#L146-L149

We can further adapt that to our needs.

mpollmeier commented:
> We can further adapt that to our needs.

You're much more familiar with mill than I am - can you please setup the build for a few scala versions (say _2.13 and _3) and I add the missing code to make it work?

lefou commented:
> > We can further adapt that to our needs.
> 
> You're much more familiar with mill than I am - can you please setup the build for a few scala versions (say _2.13 and _3) and I add the missing code to make it work?

I added some source directories that support lower or upper Scala version bound. This is based on `ZincWorkerUtil.versionRanges`:

https://github.com/com-lihaoyi/mill/blob/7eb39a83974122bfa18892454e1b5aee7177cfca/scalalib/api/src/mill/scalalib/api/ZincWorkerUtil.scala#L128-L136

If you also need exact versioned source dirs, just include `super.sources()` in your list.

mpollmeier commented:
Thank you, that worked!

I tried to split up the functionality from `object temp` into `src` (for the universally working parts), `src-2.13+` (for whatever relies on `Using`) and `src-2.12-` (with a simpler try-catch based implementation). Unfortunately I ran into all sorts of problems... 
There obviously can't be multiple `object temp`, so that didn' work. I then tried to follow the `object foo extends Function1[Path => A, A]` pattern, but I didn't find a way to get that type parameter `A` to work. 

Finally I reverted to a much simpler alternative: since it's concise and self-contained, I just copied `scala.util.Using` into our build, and changed the package to `os.util` to avoid any clashes in dowstream projects. 

That should work in theory, albeit on my local machine I'm running into some strange (and seemingy unrelated) `[error] Unwanted cyclic dependency` errors for scala 2.13 and below. Let's see what CI thinks about it.

mpollmeier commented:
About those CI failures - I wasn't convinced first, but they're actually legit... 

For some reason, calling `os.remove.all(this)`:
```scala
class TempPath private[os] (wrapped: java.nio.file.Path)
  extends Path(wrapped) with AutoCloseable {
  override def close(): Unit = os.remove.all(this)
}
```
fails in Scala 2, with rather obscure compiler errors:
```
[error] Unwanted cyclic dependency
[info] symbol: object all
[info] symbol: value up
[info] More dependencies at lines 394 394 71 87 71 152 152 87
[info] /home/mp/Projects/os-lib/os/src-jvm/package.scala:13:34:
[info]   def resource(implicit resRoot: ResourceRoot = Thread.currentThread().getContextClassLoader) = {
[info]                                  ^
[info] symbol: trait ResourceRoot
[info] More dependencies at lines 14 14
[info] /home/mp/Projects/os-lib/os/src-jvm/ResourcePath.scala:19:49:
[info]     extends BasePathImpl with ReadablePath with SegmentedPath {
[info]                                                 ^
[info] symbol: trait SegmentedPath
[info] More dependencies at lines 44 36 18 31 44 18 18 31
[error] 6 errors found
```

I fixed that by reimplementing `deleteRecursively`, and then uglifying it so that Scala 2.11 handles it as well...
Also shortened the line lengths so that `test.os.ExampleTests` is happy. 
For me locally it's all green now:
```
./mill -i -k __.clean
./mill -i -k __.test
```
Can you approve the PR run once more?

lefou reviewed (commented):
I'm not that happy with coyping the whole `scala.util.Using` into this project. We actually only want a rough replacement for `Using.resource` for internal use in case it is not provided (Scala 2.12-), which can be implemented for older Scala versions with a try-finally block. (For that we can forward the implementation of `withTempFile` to a version specific class/object.) We should not provide a full user API for convenient resource management.

Scala source code also has a different license (Apache-2) than os-lib (MIT), so copying without any additional attribution is a no go.

SethTisue commented:
fwiw, scala-collection-compat has a backport of `Using`: https://github.com/scala/scala-collection-compat/pull/319

mpollmeier commented:
Sorry about the silence! I thought it was ok because the attribution (copyright, license) are in the file, but yes I agree, it's better to have less code either way. 

Anyway, moving forward: I deleted the copied `Using.scala` and instead pull in scala-collection-compat, only for Scala 2.11 and 2.12. 
This PR is now also rebased onto latest `main`. 
`./mill -i -k __.test` is green for me locally.

mpollmeier commented:
Hmm, we seem to have some flakey tests in there, parts of `test.os.ExampleTests` fail from time to time. This doesn't seem related to this PR, the  same happens with `main` for me locally, and apparently also on CI for other builds, e.g. https://github.com/com-lihaoyi/os-lib/actions/runs/4352680832/jobs/7605762397

Any idea what's going on?

mpollmeier commented:
I'm locally running `./mill -i -k __.test` on the `main` branch in a loop and the results are very flakey:
```
grep target build-output.txt | grep ' failed$'
1 targets failed
2 targets failed
1 targets failed
1 targets failed
3 targets failed
2 targets failed
3 targets failed
1 targets failed
1 targets failed
1 targets failed
2 targets failed
1 targets failed
1 targets failed
3 targets failed
2 targets failed
2 targets failed
1 targets failed
3 targets failed
1 targets failed
1 targets failed
2 targets failed
4 targets failed
2 targets failed
1 targets failed
1 targets failed
1 targets failed
2 targets failed
1 targets failed
1 targets failed
2 targets failed
1 targets failed
3 targets failed
2 targets failed
3 targets failed
2 targets failed
1 targets failed
3 targets failed
```

mpollmeier commented:
Since this flakeyness is happening on `main`, I was curious and tested past tags as well. Even if I go back to `0.4.0` it's flakey - at least for me locally, tested on linux with openjdk 11.0.18. 

Now I'm wondering: is it just me or what's going on?

lefou commented:
@mpollmeier Thanks for your analysis. Yeah, the test suite seems flaky. Sometimes it's hard to test filesystem stuff on OSes you don't have, and especially filesystems tend to behave strange/differently in virtualized environments. 

About this PR, I hope I can review/edit/merge it soon. Sorry for the delay.

Flowdalic commented on os/src/TempOps.scala:
```diff
@@ -67,7 +79,59 @@ object temp {
+      dir = dir,
+      prefix = prefix,
+      suffix = suffix,
+      deleteOnExit = false, // TempFile.close() deletes it, no need to register with JVM
```
```suggestion
      deleteOnExit = false, // TempPath.close() deletes it, no need to register with JVM
```

Flowdalic commented on os/src/TempOps.scala:
```diff
@@ -67,7 +79,59 @@ object temp {
+    Using.resource(os.temp.dir(
+      dir = dir,
+      prefix = prefix,
+      deleteOnExit = false, // TempFile.close() deletes it, no need to register with JVM
```
```suggestion
      deleteOnExit = false, // TempPath.close() deletes it, no need to register with JVM
```

SethTisue commented:
CI seems stuck?
